### DISCUSSION TITLE
Understand cache and cache.resolve

### DISCUSSION THREAD

**Discussion by pattruong** [urql-graphql/urql#1238] [Q&A] [answered]
Hi I'm going through the documentation on normalized cache, particularly the [cache parameter](https://formidable.com/open-source/urql/docs/graphcache/computed-queries/#cache-parameter) in the resolver. I'm finding the behavior of the `cache` parameter to be a bit strange and if I could get any insight into how the cache works better that would be great 😃. 

I have a small example code where I'm making a query to fetch a list of things. There is a main type called `recipe` and then a nested type `barista`. 

**Query for list of recipes:**
```gql
query RECIPES {
  recipe {
    id
    brew_type
    water_temp
    barista {
      id
      display_name
    }
  }
}
```

From there I wanted to query a single `recipe` using an id. I wanted to test that this would be served from the cache instead of the network so I added a query resolver as well.

**Query for single recipe:**
```gql
query($id: Int!) {
  recipe_by_pk(id: $id) {
    id
    brew_type
    water_temp
    barista {
      id
      display_name
    }
  }
}
```

**`cacheExcange`:**

```js
{
  // ...client config
  exchanges: [
    devtoolsExchange,
    dedupExchange,
    cacheExchange({
      resolvers: {
        Query: {
          recipe_by_pk: (parent, args, cache, info) => {
            console.log(parent);
            const barista = cache.resolve(parent, "barista");
            const barista2 = cache.resolve(info.parentKey, "barista");
            console.log({ barista, barista2 });
            return {
              __typename: "recipe",
              id: args.id,
            };
          },
        },
      },
    }),
    fetchExchange,
  ],
}
```

Here I'm testing `cache.resolve` just to better understand how subfields are extracted. I noticed however, that none of my calls to resolve `"barista"` return a non-null value though. 

I saw this issue https://github.com/FormidableLabs/urql/issues/1207, and confirmed I am using the latest version exchange `"@urql/exchange-graphcache": "^3.3.3"`, but still seeing just `null` values in the log.

**Screenshot of log:**
<img width="1792" alt="Screen Shot 2020-12-16 at 11 55 44 AM" src="https://user-images.githubusercontent.com/6396042/102401486-4202f100-3f98-11eb-93de-1300866e1f45.png">
- You can see the `parent` parameter is populated correctly (I think it's a valid value?)
- However, even though `parent` parameter is populated with a `barista` object, it is not resolved

**Demo workflow**
- I have a component `App` that runs a `useQuery` to load all recipes
- On clicking `Select` on any of these recipes, another component `<Recipe id={id} />` gets mounted with the id of the recipe selected (pink background area)
- `Recipe` mounts and runs the `useQuery` for a single recipe (query - `recipe_by_pk`)

Overall, it works, I can confirm in urql devtools that it is getting a cache hit on running the single recipe query, I'm just a bit lost on some of the magic behind it as I'm not seeing `cache.resolve` work as I thought it would. Thanks for any insight!

**Comment by kitten** [comment, marked as answer]
Hiya 👋 I think you just looked at the incorrect field.
Notice that the parameter is called `parent` since it gets back to the parent entity and your resolver is on `Query.recipe_by_pk`.
So that means that when you call `cache.resolve(parent, ...)` the `parent` here is pointing to the `Query` entity not a recipe.

You'd get the result you expect if you added the resolver and the console logged `cache.resolve` call for `Recipe.barista` rather than `Query[field]`.

That being said, your resolver on `Query.recipe_by_pk` is returning the right thing already and the `barista` field won't need a resolver (just mentioning this to avoid confusion)

**Reply by pattruong** [reply]
@kitten Got it regarding me having the right solution.

> You'd get the result you expect if you added the resolver and the console logged cache.resolve call for Recipe.barista rather than Query[field].

Just to clarify on this part, by `Recipe.barista` you are referring to something like this?:
```js
cacheExchange({
  resolvers: {
    recipe: {
      barista: (parent, args, cache, info) => {
        const stringedKey = cache.resolve(parent, "barista");
        const id = cache.resolve(stringedKey, "id");
        const display_name = cache.resolve(stringedKey, "display_name");
        return { id, display_name };
      },
    },
  },
}),
```
In which case, my `cache.resolve` will now have `parent` as the individual `recipe` request. Logging these `cache.resolve` shows that this works as I expected, but now breaks my recipe query result causing `data` to be `null`.

**Recipe query**
```gql
query RECIPES {
  recipe {
    id
    brew_type
    water_temp
    barista {
      id
      display_name
    }
  }
}
```

I assume this is due to something with returning the `barista` field? Based on the docs, I know this method of writing a type resolver was suggested for just modifying a single field like changing a date format rather than a whole type field - e.g. `barista.id` & `barista.display_name` - so this may completely be a weird implementation just for exploratory purposes.

Overall, the cache behavior implementation is making more sense and thank you for helping me understand it 😊

**Reply by kitten** [reply]
Yea, that's correct if the type name matches and you're seeing `cache.resolve` succeed now.

What I meant prior is that it's unnecessary to tell Graphcache how to resolve `Recipe.barista` manually 😅 This will already happen automatically. In normalised caching entities are kept by keys, so by default your Recipe will be stored by `"Recipe:" + id`, which means that the "link" to `barista` will be known already by default, no matter _how_ Graphcache got to the "Recipe" entity.

To explain why your custom resolver didn't work, that's because you're returning no `__typename` field in your snippet, which should also trigger a warning in the console.

**Reply by kitten** [reply]
Oh, also to explain how returning a partial object from a resolver works internally: Graphcache will see your resolver and will drop its default behaviour, which would usually simply look up entries in the normalised cache for that field (so in this case the `barista` link for the "Recipe" entity)

It then looks at what you're returning. It already knows that it's expecting a link since you've got a selection set on `barista`. If it sees an object it first attempts to turn your object back into a key. Since a key is composed of a `__typename` field and `id` or `_id`, it looks a those fields (unless the `keys` config overrides this)

Once it has the key it can continue resolving data from the normalised cache automatically, because with the key it can continue reading from its cached data. If the object contained additional fields (like `display_name` in your example) it prefers those fields over its own data.

**Reply by pattruong** [reply]
@kitten thanks for the thorough explanation!
